test/src/us/kbase/kbasefeaturevalues/test/KBaseFeatureValuesServerTest.java

```diff
@@ -329,70 +343,94 @@ public void testPyScikitKMeans() throws Exception {
                 token, getContext());
-        ObjectData res = wscl.getObjects(Arrays.asList(new ObjectIdentity().withWorkspace(testWsName)
+        ObjectData res = getWsClient().getObjects(Arrays.asList(new ObjectIdentity().withWorkspace(testWsName)
                 .withName(clustObj1Name))).get(0);
+
+        //test if returned id matches id returned by ws saveobjects
+        Assert.assertEquals(getInfoEKN.get(0).getE7()  + "/" + getInfoEKN.get(0).getE1() + "/" +  getInfoEKN.get(0).getE5(), getId);
+
         FeatureClusters clSet = res.getData().asClassInstance(FeatureClusters.class);
         checkKMeansForSample(clSet);
     }
-    
-    @Test
+
     public void testCorrectMatrix() throws Exception {
         String testWsName = getWsName();
         String sourceMatrixId = "notcorrected_matrix.1";
+
+        FloatMatrix2D testmat = getSampleMatrix();
         ExpressionMatrix data = new ExpressionMatrix().withType("log-ratio").withScale("1.0")
-                .withData(getSampleMatrix());
+                .withData(testmat);
+        //introduce null value for imputation test
         data.getData().getValues().get(0).set(0, null);
         Assert.assertEquals(1, getNullCount(data.getData()));
-        getWsClient().saveObjects(new SaveObjectsParams().withWorkspace(testWsName).withObjects(Arrays.asList(
+        List<Tuple11<Long, String, String, String, Long, String, Long, String, String, Long, Map<String, String>>> infoCM = getWsClient().saveObjects(new SaveObjectsParams().withWorkspace(testWsName).withObjects(Arrays.asList(
                 new ObjectSaveData().withName(sourceMatrixId).withType("KBaseFeatureValues.ExpressionMatrix")
-                .withData(new UObject(data)))));
+                        .withData(new UObject(data)))));
         String targetMatrixId = "corrected_matrix.1";
-        impl.correctMatrix(new CorrectMatrixParams().withInputData(
+        String getId = impl.correctMatrix(new CorrectMatrixParams().withInputData(
                 testWsName + "/" + sourceMatrixId).withOutWorkspace(testWsName)
                 .withOutMatrixId(targetMatrixId).withTransformType("missing"), token, getContext());
+
         ExpressionMatrix matrix = getWsClient().getObjects(Arrays.asList(
                 new ObjectIdentity().withWorkspace(testWsName).withName(targetMatrixId)))
                 .get(0).getData().asClassInstance(ExpressionMatrix.class);
+
+        //test matrix imputation
         Assert.assertEquals(0, getNullCount(matrix.getData()));
-        Assert.assertEquals(0.325, (double)matrix.getData().getValues().get(0).get(0), 1e-10);
+        Assert.assertEquals(0.325, (double) matrix.getData().getValues().get(0).get(0), 1e-10);
+
+        //reload with returned object id
+        ExpressionMatrix res = loadfromWs(getId).asClassInstance(ExpressionMatrix.class);
+        //reset from null in data before imputation
+        data.getData().getValues().get(0).set(0, res.getData().getValues().get(0).get(0));
```

Comment:
Can you explain what's going on in this line? Maybe expand the comment. It's not clear to me what's going on.

Reply:
this is testing the matrix values globally, but have to change one in test matrix to account for data imputation

actually this is already covered under testSaveMatrix() -- but can't hurt

Reply:
So how does the test ensure that the imputed value is correct?

Reply:
this test does not but a few lines above:

Assert.assertEquals(0.325, (double) matrix.getData().getValues().get(0).get(0), 1e-10);

Reply:
Ok, i'm a little confused then - the test pulls the same object from the workspace twice, and runs different tests on the two instances of the same object?

Reply:
Yes, but it pulls the object using two different ways 1) ws and object name specified in the code and 2) ws reference returned by the method. It may be confusing at first but I we do want to test both the returned object contents as well as the ws ref for that object.

Reply:
>  It may be confusing at first but I we do want to test both the returned object contents as well as the ws ref for that object.

But you can do that by just pulling once via the returned ref, surely?

Reply:
sure, you could pull once but then test is combined, isn't it better to isolate failures?

but I just combined into one load...